### ISSUE TITLE
Migration to PyTorch core

### ISSUE THREAD

**Issue by jansel** [pytorch/torchdynamo#1588]
We have merged `github.com/pytorch/torchdynamo` into `github.com/pytorch/pytorch`.  Being in a separate repo has allowed us to move fast, but it creates some installation roadblocks for users and version pinning has been a growing headache as the project scales.

- `import torchdynamo` is now `import torch._dynamo`
- `import torchinductor` is now `import torch._inductor`
- The standalone C extensions are added to `torch._C`
- `torchdynamo/benchmarks` is now `pytorch/benchmarks/dynamo`
- `torchdynamo/test/dynamo` is now `pytorch/test/dynamo`
- `torchdynamo/test/inductor` is now `pytorch/test/inductor` 

# Instructions for migrating your PR from `pytorch/torchdynamo` to `pytorch/pytorch`

1) Merge/rebase your PR to the correct hash of TorchDynamo: https://github.com/pytorch/torchdynamo/commit/6380959be21851bfda99424392cc08fda29d073d
```
cd ../torchdynamo
git pull origin 6380959be21851bfda99424392cc08fda29d073d
```

2) Checkout a new branch with the correct version of PyTorch: https://github.com/pytorch/pytorch/commit/054a2fd6c2fc361796663eed4772368c287d6c83
(the two versions must match so only your changes are copied over):
```
cd ../pytorch
git fetch origin master 
git checkout -b <your_new_branch_name> 054a2fd6c2fc361796663eed4772368c287d6c83
```

3) Run the rsync+auto-format script
```
bash ../torchdynamo/copy_to_core.sh
```

4) Commit your changes, rebase to the latest version of PyTorch, and create a pull request as normal for the PyTorch repo.

If you are copying files by hand, make sure to run `make setup_lint && lintrunner -a` to switch to the new auto-formatting.

# The Plan

## Phase 1 (complete)

In the first phase we will move the code over to core, which consists of these two PRs:
https://github.com/pytorch/torchdynamo/pull/1538
https://github.com/pytorch/pytorch/pull/86461

In this phase (now complete), the dynamo source of truth remained `pytorch/torchdynamo`.  Any changes will be copied over by the `copy_to_core.sh` script which we will rerun at the end of phase one.

After these initial PRs land, dynamo will be working in core -- but most of the testing will be disabled.  The initial PR will have testing disable, but then we will enable tests in smaller followup PRs.

## Phase 2 (going on now)

At this switchover point, the source of truth will become PyTorch core.  All open PR's will be closed and asked to resubmit to `pytorch/pytorch`.  We will replace `torchdynamo` in this repo with `from torch._dynamo import *`, and same for `torchinductor`.

The criteria for reaching this point is the PRs above land in core, sync to fbcode (to check for reverts), and basic testing infra is setup.

We will continue using this project for github issues.

# Followup Tasks

- [x] Land initial migration in core (this blocks most other items) @jansel
- [x] Land dynamo size of that migration with `copy_to_core.sh` script @jansel
- [x] Enable all unit tests in core @jansel
- [x] Add smaller deps such as Sympy/Filelock and to PyTorch CI
- [x] `test_recompile_ux.py` has a bug where it only works in `pytest` and not using PyTorch's test infra  @wconstab 
- [x] Add Triton to PyTorch CI https://github.com/pytorch/pytorch/pull/86988
- [x] Update torchdynamo tests in core to use `torch._dynamo` and stop installing `torchdynamo` https://github.com/pytorch/pytorch/pull/87039
- [ ] Figure out how to include triton/etc properly for end users installing nightlies.  Perhaps vend our own version of it in the short term.  @malfet 
- [x] Update performance dashboards to use the new system.  @anijain2305 
- [ ] Move docs over to pytorch core (these won't be covered in the initial move)
- [x] Update internal fbcode buck target and internal users to use the new packages, remove the old one
- [ ] Port over our circleci integration tests, we may need to convert these to github actions in the process.  Since these are expensive, we may want to gate them on a tag.  @desertfire 
- [ ] Fix followup issues uncovered by @albanD, @voznesenskym

I am sure there are other things I am forgetting.  Please edit this or comment with suggestions.

**Comment by wconstab** [comment]
what's the best way to run/repro the test_recompile_ux bug?

**Comment by jansel** [comment]
> what's the best way to run/repro the test_recompile_ux bug?
1) Checkout https://github.com/pytorch/torchdynamo/pull/1538
2) Uncomment these lines:
https://github.com/pytorch/torchdynamo/blob/151fc2517c8e9d7ae9c257476190c7dadca87c69/test/dynamo/test_recompile_ux.py#L201-L204
3) `python test/dynamo/test_recompile_ux.py` should fail (while pytest passes)

**Comment by jansel** [comment]
> Windows test enablement
https://github.com/pytorch/torchdynamo/issues/64
> Py3.11 eventual support
https://github.com/pytorch/torchdynamo/issues/106

**Comment by malfet** [comment]
Consider splitting https://github.com/pytorch/pytorch/pull/86461  into two parts: native plugins and pure python code.

**Comment by wconstab** [comment]
>what's the best way to run/repro the test_recompile_ux bug?

@jansel is it important for  torchdynamo.git/test/dynamo/test_recompile_ux.py to pass?  I did scratch my head debugging this and was not able to figure out why the logger config is not working there.  However, pytorch.git/test/dynamo/test_recompile_ux.py does pass for me.  Isn't that all that matters at this point?

**Comment by jansel** [comment]
I think I disabled the test, so it might be passing because of that.  If it is too hard I am ok just disabling it.

**Comment by wconstab** [comment]
I double checked my tmux scrollback and i did run it with `-v` and see that the tests really ran.  (`Ran 6 tests in 2.375s`) -- i'm pretty sure something is messed up with the version of the test on the torchdynamo repo i think things are ok on the pytorch side if you just uncomment that block

```
-# if __name__ == "__main__":
-#     from torch._dynamo.testing import run_tests
-#     run_tests()
```

I am also ok with it being disabled transiently, so if you find that enabling it doesn't work just disable and file an issue on me, i'll debug it once the transition has settled.

**Comment by malfet** [comment]
Do we need followup task for fixing internal build config to point to the "right" version for torchdynamo?

**Comment by jansel** [comment]
> [ ] Update internal fbcode buck target and internal users to use the new packages, remove the old one

That is what I meant by "Update internal fbcode buck target and internal users to use the new packages, remove the old one" in the list above

**Comment by jansel** [comment]
The initial migration has landed, so most other tasks should be unblocked.

**Comment by Chillee** [comment]
Will the old import still work or do any users of torchdynamo need to change their imports too?

**Comment by jansel** [comment]
> Will the old import still work or do any users of torchdynamo need to change their imports too?

They should.  I plan to replace the old locations with a file like:

torchdynamo.py
```
from torch._dynamo import *  # noqa
```

I plan to wait a bit before doing that, as it will create a ton of merge conflicts for people porting PRs over if the `git pull` by mistake.

**Comment by jansel** [comment]
> Hey Jason, after this move, what is the "TorchDynamo homepage"? -- I would previously point folks at torch/torchdynamo:README.md but I can't find an analogous file in pytorch/pytorch now.

We still need to move all the docs over.  Eventually it will be a page on https://pytorch.org/docs/ -- but for now you can keep linking here.

**Comment by wconstab** [comment]
what's the plan/status of unit tests? I wanted to update/re-enable test_distributed.py for example, but not sure if I should try to do that yet (or if so, in which CI I should expect it to run).

**Comment by jansel** [comment]
If you uncomment this, it should re-enable distributed tests:
https://github.com/pytorch/pytorch/blob/4814270708cb6141c1fb6202f883c084c71290b4/test/dynamo/test_distributed.py#L285-L287

Our unittests are standard pytorch tests and run in most test configurations.

**Comment by wconstab** [comment]
I was getting `ImportError: cannot import name 'run_tests' from 'torch._dynamo.testing' (/opt/conda/lib/python3.7/site-packages/torch/_dynamo/testing.py)` -- should i be removing the _dynamo.testing references and replacing with the standard pytorch ones?

**Comment by jansel** [comment]
@wconstab ah no, it just got renamed to `from 'torch._dynamo.test_case import run_tests`.

Needed to put it in a separate file due to `import torch.testing.*` having global side effects.
